src/grappler/ngraph_optimizer.cc

```diff
@@ -0,0 +1,188 @@
+    return Status::OK();
+  }
+
+  // Do variable capture then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(CaptureVariables(&graph, skip_these_nodes));
+  if (DumpCapturedGraphs()) {
+    DumpGraphs(graph, idx, "captured", "Graph With Variables Captured");
+  }
+
+  //
+  // Encapsulation: Part that rewrites the graph for nGraph operation.
+  //
+  // The part has several phases, each executed in sequence:
+  //
+  //   1. Marking [ngraph_mark_for_clustering.cc]
+  //   2. Cluster Assignment [ngraph_assign_clusters.cc]
+  //   3. Cluster Deassignment [ngraph_deassign_clusters.cc]
+  //   4. Cluster Encapsulation [ngraph_encapsulate_clusters.cc] - currently
+  //      part of the ngraph_rewrite_pass.cc to be executed after POST_REWRITE
+  //
+  // Between phases, graph dumps (in both .dot and .pbtxt format) may be
+  // requested by setting the following environment variables:
+  //
+  //   NGRAPH_TF_DUMP_UNMARKED_GRAPHS=1      dumps graphs before phase 1
+  //   NGRAPH_TF_DUMP_MARKED_GRAPHS=1        dumps graphs after phase 1
+  //   NGRAPH_TF_DUMP_CLUSTERED_GRAPHS=1     dumps graphs after phase 2
+  //   NGRAPH_TF_DUMP_DECLUSTERED_GRAPHS=1   dumps graphs after phase 3
+  //   NGRAPH_TF_DUMP_ENCAPSULATED_GRAPHS=1  dumps graphs after phase 4
+  //   NGRAPH_TF_DUMP_GRAPHS=1               all of the above
+  //
+
+  // If requested, dump unmarked graphs.
+  if (DumpUnmarkedGraphs()) {
+    DumpGraphs(graph, idx, "unmarked", "Unmarked Graph");
+  }
+
+  // If ngraph is disabled via ngraph_bridge api or NGRAPH_TF_DISABLE is set
+  // we will not do anything; all subsequent
+  // passes become a no-op.
+  if (config::IsEnabled() == false ||
+      std::getenv("NGRAPH_TF_DISABLE") != nullptr) {
+    return Status::OK();
+  }
+
+  // 1. Mark for clustering then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(MarkForClustering(&graph, skip_these_nodes));
+  if (DumpMarkedGraphs()) {
+    DumpGraphs(graph, idx, "marked", "Graph Marked for Clustering");
+  }
+
+  // 2. Assign clusters then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(AssignClusters(&graph));
+  if (DumpClusteredGraphs()) {
+    DumpGraphs(graph, idx, "clustered", "Graph with Clusters Assigned");
+  }
+
+  // 3. Deassign trivial clusters then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(DeassignClusters(&graph));
+  if (DumpDeclusteredGraphs()) {
+    DumpGraphs(graph, idx, "declustered",
+               "Graph with Trivial Clusters De-Assigned");
+  }
+
+  // 4. Encapsulate clusters then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(EncapsulateClusters(&graph));
+  if (DumpEncapsulatedGraphs()) {
+    DumpGraphs(graph, idx, "encapsulated", "Graph with Clusters Encapsulated");
+  }
+
+  // Rewrite for tracking then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(RewriteForTracking(&graph));
+  if (DumpTrackedGraphs()) {
+    DumpGraphs(graph, idx, "tracked",
+               "Graph with Variables Rewritten for Tracking");
+  }
```

Comment:
This core logic of `MarkForClustering`-->`AssignClusters`-->`DeassignClusters`-->`EncapsulateClusters`-->`RewriteForTracking` is also present in `class NGraphVariableCapturePass : public NGraphRewritePass`. The functions in use here are independent of the classes and might be moved out to some `core_rewrite` function, which is used by both.